### PR TITLE
planner: rewrite to make FULL JOIN executable

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### What is changed and how it works?

rewrite `t1 full join t2 on COND` to `t1 left join t2 on COND union all select null, null, ... null, t2.* from t2 where not exists(select 1 from t1 where COND)`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by elsa0520 — DISMISSED** [review]
LGTM

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
...
+		rFullNames = right.fullNames
+	} else {
+		rFullSchema = rChild1.Schema()
+		rFullNames = rChild1.OutputNames()
+	}
+	if joinNode.Tp == ast.RightJoin {
+		// Make sure lFullSchema means outer full schema and rFullSchema means inner full schema.
+		lFullSchema, rFullSchema = rFullSchema, lFullSchema
+		lFullNames, rFullNames = rFullNames, lFullNames
+	}
```
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:913]: This case is impossible now.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
...
+	for _, rName := range rFullNames {
+		name := *rName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+
+	// Clear NotNull flag for the inner side schema if it's an outer join.
+	if joinNode.Tp == ast.LeftJoin || joinNode.Tp == ast.RightJoin {
+		resetNotNullFlag(leftJoinPlan.fullSchema, 0, leftJoinPlan.fullSchema.Len())
+	}
```
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:930]: The `joinNode.Tp` must be `FullJoin`, but we need to treat the logical join as a `LeftJoin` here.

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
...
+	handleMap1 = b.handleHelper.popMap()
+	handleMap2 = b.handleHelper.popMap()
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
+
+	antiJoinPlan := LogicalJoin{StraightJoin: joinNode.StraightJoin || b.inStraightJoin}.Init(b.ctx, b.getSelectOffset())
+	antiJoinPlan.SetChildren(rChild2, lChild2)
+	antiJoinPlan.SetSchema(expression.MergeSchema(rChild2.Schema(), lChild2.Schema()))
+	antiJoinPlan.names = make([]*types.FieldName, rChild2.Schema().Len()+lChild2.Schema().Len())
+	copy(antiJoinPlan.names, rChild2.OutputNames())
+	copy(antiJoinPlan.names[rChild2.Schema().Len():], lChild2.OutputNames())
```
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:963]: Looks like we can remove these code.

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
...
+	projExprs := make([]expression.Expression, lChild2.Schema().Len()+rChild2.Schema().Len())
+	copy(projExprs[lChild2.Schema().Len():], expression.Column2Exprs(rChild2.Schema().Columns))
+	for i := 0; i < lChild2.Schema().Len(); i++ {
+		projExprs[i] = expression.NewNull()
+	}
+	proj := LogicalProjection{
+		Exprs: projExprs,
+	}.Init(b.ctx, b.getSelectOffset())
+	proj.SetChildren(antiJoinPlan)
+	proj.SetSchema(expression.MergeSchema(lChild2.Schema(), rChild2.Schema()))
```
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:990]: We didn't set output names for the `proj`.

Comment (winoros) [reply planner/core/logical_plan_builder.go:990]: We just need to set it for the top layer of the rewrite.

File: planner/core/logical_plan_builder.go (hunk 3 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go:963]: Yes

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
...
+	if err != nil {
+		return nil, err
+	}
+	rChild2, err := b.buildResultSetNode(ctx, joinNode.Right, false)
+	if err != nil {
+		return nil, err
+	}
+	handleMap1 = b.handleHelper.popMap()
+	handleMap2 = b.handleHelper.popMap()
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
```
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:956]: Does the order of `handleMap1` and `handleMap2` matter here?

File: planner/core/logical_plan_builder.go (hunk 4 above)
Comment (time-and-fate) [inline planner/core/logical_plan_builder.go:990]: Looks like we should also reset the NotNULL flag for the `lChild2` schema in the projection schema here.

File: planner/core/logical_plan_builder.go (hunk 5 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go:956]: No

File: planner/core/logical_plan_builder.go (hunk 4 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go:990]: I'm going to reset the flags for all columns.

File: planner/core/logical_plan_builder.go (hunk 3 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go:963]: The following constructing ON will need the output names to find columns.

**Review by fixdb — APPROVED** [review]
+1

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**